### DISCUSSION TITLE
How to handle unauthenticated case when using user profile method?

### DISCUSSION THREAD

**Discussion by victor-ivanov** [aws/aws-sdk-net#3535] [Q&A] [answered]
We are using `AWSSDK.Extensions.NETCore.Setup` package to authenticate with AWS services. 

```
var awsOptions = new AWSOptions
{
    Region = RegionEndpoint.USEast1,
    Profile = "some-profile-name",
};
var client = awsOptions.CreateServiceClient<IAmazonS3>();
var buckets = await client.ListBucketsAsync();
```

It throws following error if user is not authenticated via aws cli
```
System.ArgumentNullException: Value cannot be null. (Parameter 'Options property cannot be empty: ClientName')

Stack Trace: 
  SSOTokenManager.GenerateNewTokenAsync(SSOTokenManagerGetTokenOptions options, CancellationToken cancellationToken) line 623
  SSOTokenManager.GetTokenAsync(SSOTokenManagerGetTokenOptions options, CancellationToken cancellationToken) line 526
  SSOAWSCredentials.GetSsoCredentialsAsync(ICoreAmazonSSO sso) line 230
  SSOAWSCredentials.GenerateNewCredentialsAsync() line 153
  RefreshingAWSCredentials.GetCredentialsAsync() line 154
  CredentialsRetriever.InvokeAsync[T](IExecutionContext executionContext) line 99
  RetryHandler.InvokeAsync[T](IExecutionContext executionContext) line 135
  RetryHandler.InvokeAsync[T](IExecutionContext executionContext) line 150
  CallbackHandler.InvokeAsync[T](IExecutionContext executionContext) line 61
  CallbackHandler.InvokeAsync[T](IExecutionContext executionContext) line 61
  AmazonS3ExceptionHandler.InvokeAsync[T](IExecutionContext executionContext) line 57
  ErrorCallbackHandler.InvokeAsync[T](IExecutionContext executionContext) line 53
  MetricsHandler.InvokeAsync[T](IExecutionContext executionContext) line 92
```

Is it possible to make this code throw some kind of `UnauthenticatedException`  if user is not logged in into cli? Or should I handle this specific `System.ArgumentNullException` as no auth case?

**Comment by ashishdhingra** [comment, marked as answer]
@victor-ivanov Good afternoon. Thanks for starting the discussion. Looks like you are using SSO profile. To use SSO profile, one needs to be authenticated via `aws sso login` and AWS .NET SDK takes care of the rest. In your use case, since you didn't execute AWS CLI command `aws sso login`, it failed while validating `ClientName` [here](https://github.com/aws/aws-sdk-net/blob/b57ca5ff76e1698f48d6b9fedc6bf5a0545e429b/sdk/src/Core/Amazon.Runtime/Credentials/Internal/_bcl45%2Bnetstandard/SSOTokenManager.cs#L354). I do not see an option other than to catch `System.ArgumentNullException` and look out for specific exception message (which I agree is not scalable).

**EDIT:** I have reviewed this with the team and created an internal backlog item to use better exception classes.

Thanks,
Ashish

**Reply by ashishdhingra** [reply]
This has been addressed in [AWSSDK.Core version 4.0.0-preview.12](https://www.nuget.org/packages/AWSSDK.Core/4.0.0-preview.12)
